libpromises/rlist.c

```diff
@@ -1206,35 +1204,35 @@ void RlistWrite(Writer *writer, const Rlist *list)
     WriterWriteChar(writer, '}');
 }
 
-/* Note: only single quotes are escaped, as they are used in RlistWrite to
-   delimit strings. If double quotes would be escaped, they would be mangled by
-   RlistParseShown */
-
-static void ScalarWrite(Writer *w, const char *s)
+void ScalarWrite(Writer *writer, const char *s, bool dontQuote)
 {
+    if(!dontQuote)
+        WriterWriteChar(writer, '"');
     for (; *s; s++)
     {
-        if (*s == '\'')
+        if (*s == '"')
         {
-            WriterWriteChar(w, '\\');
+            WriterWriteChar(writer, '\\');
         }
-        WriterWriteChar(w, *s);
+        WriterWriteChar(writer, *s);
     }
+    if(!dontQuote)
+        WriterWriteChar(writer, '"');
 }
 
-void RvalWrite(Writer *writer, Rval rval)
+void RvalWrite(Writer *writer, Rval rval, bool dontQuote)
 {
-    RvalWriteParts(writer, rval.item, rval.type);
+    RvalWriteParts(writer, rval.item, rval.type, dontQuote);
 }
 
 char *RvalToString(Rval rval)
 {
     Writer *w = StringWriter();
-    RvalWrite(w, rval);
+    RvalWrite(w, rval, true);
     return StringWriterClose(w);
 }
 
-void RvalWriteParts(Writer *writer, const void* item, RvalType type)
+void RvalWriteParts(Writer *writer, const void* item, RvalType type, bool dontQuote)
```

Comment:
You're right, this function is too long to duplicate. But in fact, it's not used anywhere besides this file, so keep the name and the `quote` argument, but make it `static` instead, and remove it from `rlist.h` (possibly including the declaration at the top of this file instead).